### ISSUE TITLE
gotchya! including `remake.yml` can really slow down complex pipeline task tables

### ISSUE THREAD

**Issue by lindsayplatt** [USGS-R/scipiper#150]
**The issue:** I have been working in https://github.com/USGS-R/lake-temperature-out and running the `3_summarize` part of the pipeline. So far, I have had no issue building task tables with 880 tasks that take ~8 min each in parallel on Yeti UV. I just started to build this for a task table with about 7,000 tasks and have struggled. It gets hung up prior to starting the loop, as it checked which targets in the task table needed to be rebuilt. One of the earlier suggestions (from @aappling-usgs) to fix this issue was to save indicator files rather than use objects or actual data files as targets in the task table. I implemented this approach, but without any noticeable improvements. Even trying to build just one target in this big task makefile (a la `scmake("task_target", "task_makefile.yml")`) was impossible because it would take hours before it would start the actual build (I didn't wait to see how long because 1 hour to figure out how to build a target that should take 10 minutes seemed ridiculous). 

**The current solution:** After talking with @jread-usgs, it became clear that listing the `remake.yml` in the `include:` field of the task makefile might be causing this slow down. When this is added, it is adding complexity to an already lengthy dependency tree. I was using `include: remake.yml` because there were a few object targets that were being used in the task steps. The current solution is to decouple the task makefile from the rest of the pipeline by removing `remake.yml` from `include:` and then saving the required object targets as temporary files during the task makefile creation (which get deleted once `loop_tasks` is complete). I am still testing this out right now, but it worked well when I built just one task in the huge task makefile. The single task took 8 minutes, which is what I expected (the same attempt before took longer than an hour ...). When I kicked off the full build, it took 25 minutes to get from kicking off the build to printing `### Starting loop attempt ... `. Way better than the hours I had waited previously ... sometimes my Yeti UV job would just die without any errors.

**Some concerns:** Since we are building these tasks in parallel (I am currently using 80 cores on Yeti UV), will it be an issue that we have 80 functions trying to all read in the same RDS file at once? We aren't sure yet. 

**Possible outcomes:**

1. Should we consider other code options? One idea I had was that it would be nice to add some targets to the task makefile that aren't repeated per task. This would have allowed me to add the object targets one time at the beginning of the task makefile and then repeat the rest. We could do this in a separate temporary makefile, too. These options seemed more convoluted and required more effort than switching the object targets to files.
2. Does this "gotchya" need to be documented somewhere in `scipiper`? Would a user get notified if this is a concern with how they have setup their pipeline? Perhaps some threshold of number of targets in the task table could be used to push a warning?
3. Does the approach need to be outlined in a future pipelines training: "considerations when scaling up task tables"?

**Comment by aappling-usgs** [comment]
Nice writeup, Lindsay, and good to know about this solution. In hindsight, it does make sense that adding thousands of targets that each depend on [the same] long chain of other targets could really bog down the dependency manager. It's a shame that the solution is to create copies of the data (in this case where the original dependencies were object targets), but I'm glad you've found a way to do it that should preserve data and pipeline integrity.

**Comment by aappling-usgs** [comment]
I will be interested to learn about your experience trying to read the same file in many tasks simultaneously on Yeti. I hope that cluster computing systems have better-than-laptop ability to support concurrent reads, but I don't know either. Is this question related to the `include` problem and solution that's the main topic of this issue, though? If so, how?

**Comment by aappling-usgs** [comment]
I think you've made the case in the past for having non-task-table targets within a task table yml, and I agree that sometimes that would feel cleaner, especially because in practice our task tables are also sets of tasks that we expect to run on just one computer/cluster within our analysis team (and so it would be nice to lump other closely-related tasks in with the table). If we did have that capability (or if you took the plunge on writing a wrapper / separate temporary makefile), would you intentionally decouple the object targets from a need to `include: remake.yml`, and if so, how would you do it?

**Comment by aappling-usgs** [comment]
> Perhaps some threshold of number of targets in the task table could be used to push a warning?

Hmm. Would the number of targets in the task table be the right metric? I wonder if it has as much to do with the length of the dependency chain on which all of those targets are depending (and the size of the dependencies, complexity of the functions, etc.)

**Comment by aappling-usgs** [comment]
> Does the approach need to be outlined in a future pipelines training?

I do like the idea of adding a note about task table scaling, probably to pipelines 3. I've added an issue there (https://github.com/USGS-R/ds-pipelines-3-course/issues/28).

**Comment by jread-usgs** [comment]
> Hmm. Would the number of targets in the task table be the right metric? 

Yeah, I don't think so. The reason this issue was a little surprising to me was that I'd done tests of thousands of simple tasks in a task table and didn't see any of this trouble. The issue is the complexity of the dependencies + the number of tasks (and dependencies) I assume here.

**Comment by lindsayplatt** [comment]
It is checking a lot of files when I use `include: remake.yml` because I have targets in `2_process` that unzip and save individual lake files as well. So just between the actual tmp files, and then the two targets per task in the task table, that is 21,000 files to check. That doesn't count all of the other groups of files that I built for the other models.

**Comment by lindsayplatt** [comment]
> I will be interested to learn about your experience trying to read the same file in many tasks simultaneously on Yeti. 

At first it didn't seem like an issue. I got a little greedy, though, and tried upping the cores from 80 to 150. After about 20 minutes of working on `loop_tasks`, it failed with the following error. I am not sure if it is related. I have since restarted the build.

```
Error in socketConnection("localhost", port = port, server = TRUE, blocking = TRUE,  : 
  all connections are in use
Calls: scmake ... newPSOCKnode -> socketConnection -> .handleSimpleError -> h
```

> Is this question related to the include problem and solution that's the main topic of this issue, though? If so, how?

Related only because in order to decouple the `remake.yml`, I had to convert two object targets into file targets that each task in the task plan must read. Now, rather than passing an object to each concurrent task, I am passing an filename that is read in by the each task step.

**Comment by lindsayplatt** [comment]
> If we did have that capability (or if you took the plunge on writing a wrapper / separate temporary makefile), would you intentionally decouple the object targets from a need to include: remake.yml, and if so, how would you do it?

Excellent question ... I am not really sure if this is a strategy worth pursuing since using files instead of objects appears to be working (except for the minor hiccup described in the previous comment). As I think more about it, keeping this outside of `scipiper` functions and as a strategy that you implement in your individual pipeline would be best.

In my current use case, I have an object target that is a two-column data frame. It is created by first reading in a TSV file and then subsetting to just two columns. I did this because other columns in this data set (such as source and description) have changed but the actual range values have not and I didn't want those other changes to kick off a rebuild of the thousands of lake targets. 

```
temp_range_info:
    command: read_tsv("3_summarize/in/temp_ranges_for_metrics.txt")
temp_ranges:
    command: select(temp_range_info, I('Temp_Low'), I('Temp_High'))
```

Since "3_summarize/in/temp_ranges_for_metrics.txt" is a file and not a target elsewhere in the pipeline, I could scoot both of the targets in the chunk of above into a new makefile called `task_plan_targets.yml` or something. Then, I could pass that filename in as the `include`-ed yml rather than `remake.yml`.

**Comment by aappling-usgs** [comment]
> Related only because in order to decouple the remake.yml, I had to convert two object targets into file targets that each task in the task plan must read. Now, rather than passing an object to each concurrent task, I am passing an filename that is read in by the each task step.

Gotchya, thanks. I want to clarify that in each new R session (i.e., in each Slurm job or task that calls Rscript), even if something looks like an object target, `remake`/`scipiper` is still reading an .rds file in order to reload that object into the session.

**Comment by lindsayplatt** [comment]
Ahhhh 💡 That is quite different from how I have been imagining it.
